pkg/provisioners/channel_util.go

```diff
@@ -50,17 +51,44 @@ func RemoveFinalizer(c *eventingv1alpha1.Channel, finalizerName string) {
 }
 
 func CreateK8sService(ctx context.Context, client runtimeClient.Client, c *eventingv1alpha1.Channel) (*corev1.Service, error) {
-	svcKey := types.NamespacedName{
-		Namespace: c.Namespace,
-		Name:      ChannelServiceName(c.Name),
+	getSvc := func() (*corev1.Service, error) {
+		return getK8sService(ctx, client, c)
 	}
-	return createK8sService(ctx, client, svcKey, newK8sService(c))
+	return createK8sService(ctx, client, getSvc, newK8sService(c))
 }
 
-func createK8sService(ctx context.Context, client runtimeClient.Client, svcKey types.NamespacedName, svc *corev1.Service) (*corev1.Service, error) {
-	current := &corev1.Service{}
-	err := client.Get(ctx, svcKey, current)
+func getK8sService(ctx context.Context, client runtimeClient.Client, c *eventingv1alpha1.Channel) (*corev1.Service, error) {
+	list := &corev1.ServiceList{}
+	opts := &runtimeClient.ListOptions{
+		Namespace:     c.Namespace,
+		LabelSelector: labels.SelectorFromSet(k8sServiceLabels(c)),
+		// TODO this is here because the fake client needs it. Remove this when it's no longer
+		// needed.
+		Raw: &metav1.ListOptions{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: corev1.SchemeGroupVersion.String(),
+				Kind:       "Service",
+			},
+		},
+	}
 
+	err := client.List(ctx, opts, list)
+	if err != nil {
+		return nil, err
+	}
+	for _, svc := range list.Items {
+		if metav1.IsControlledBy(&svc, c) {
+			return &svc, nil
+		}
+	}
+
+	return nil, k8serrors.NewNotFound(schema.GroupResource{}, "")
+}
+
+type getService func() (*corev1.Service, error)
```

Comment:
We call `createK8sService` in two places. Once for the general dispatcher (provisioner_util.go), and once for each `Channel` (here). This PR changes the methodology for finding the `Service` associated with each `Channel` to use a list with a label selector. It does not change the methodology for find the `Service` associated with the dispatcher (still uses `client.Get`). 

My reasoning is that no one cares what the `Channel`'s address is, it should always be read from that `Channel`'s `status.address`, so a generated name makes sense, which implies list with a label selector. I like the dispatcher having a well known name, so that we don't need to read the `ClusterChannelProvisioner` on every `Channel` reconcile (or cache it). And if it didn't have a well known name, then the 'address' of the dispatcher could change and we would need to reconcile **every** `Channel` associated with it when it changes. 

Because I want the well known name for the dispatcher, `client.Get` makes more sense than list with label selector. By passing this function into `createK8sService`, I can use identical code paths for both the dispatcher and every `Channel`, swapping out the finding of the associated `Service`.

Reply:
ack.